urlfb310a588a_alokhan_LinkStateNetworkSimulation_tgz-pJ8-reso_examples_static_routing_DemoJ8/src/reso/examples/dv_routing/DVRoutingProtocol.java

```diff
@@ -0,0 +1,201 @@
+
+import reso.common.AbstractApplication;
+import reso.common.Interface;
+import reso.common.InterfaceAttrListener;
+import reso.examples.dv_routing.DVMessage.DV;
+import reso.ip.Datagram;
+import reso.ip.IPAddress;
+import reso.ip.IPInterfaceAdapter;
+import reso.ip.IPInterfaceListener;
+import reso.ip.IPLayer;
+import reso.ip.IPLoopbackAdapter;
+import reso.ip.IPRouter;
+
+public class DVRoutingProtocol
+extends AbstractApplication
+implements IPInterfaceListener, InterfaceAttrListener {
+
+	public static final String PROTOCOL_NAME= "DV_ROUTING";
+	public static final int IP_PROTO_DV= Datagram.allocateProtocolNumber(PROTOCOL_NAME);
+	
+	private final IPLayer ip;
+	private final boolean advertise;
+	
+	private final Map<IPAddress,Map<IPAddress,DVMessage.DV>> table=
+		new HashMap<IPAddress,Map<IPAddress,DVMessage.DV>>();
+	
+	/** Constructor
+	 * 
+	 * @param router is the router that hosts the routing protocol
+	 * @param advertise specifies if this router will send DV for its own local destinations
+	 */
+	public DVRoutingProtocol(IPRouter router, boolean advertise) {
+		super(router, PROTOCOL_NAME);
+		this.ip= router.getIPLayer();
+		this.advertise= advertise;
+	}
+
+	private IPAddress getRouterID() {
+		IPAddress routerID= null;
+		for (IPInterfaceAdapter iface: ip.getInterfaces()) {
+			IPAddress addr= iface.getAddress();
+			if (routerID == null)
+				routerID= addr;
+			else if (routerID.compareTo(addr) < 0)
+				routerID= addr;
+		}
+		return routerID;
+	}
+	
+	@Override
+	public void start() throws Exception {
+		// Register listener for datagrams with DV routing messages
+		ip.addListener(IP_PROTO_DV, this);
+		
+		// Register interface attribute listeners to detect metric and status changes
+		for (IPInterfaceAdapter iface: ip.getInterfaces())
+			iface.addAttrListener(this);
+		
+		// Send initial DV
+		if (advertise) {
+			for (IPInterfaceAdapter iface: ip.getInterfaces()) {
+				if (iface instanceof IPLoopbackAdapter)
+					continue;
+				DVMessage dvm= new DVMessage();
+				//dvm.addDV(ip.getInterfaceByName("lo0").getAddress(), 0, null);
+				dvm.addDV(getRouterID(), 0, null);
+				iface.send(new Datagram(iface.getAddress(), IPAddress.BROADCAST, IP_PROTO_DV, 1, dvm), null);
+			}
+		}
+	}
+	
+	public void stop() {
+		ip.removeListener(IP_PROTO_DV, this);
+		for (IPInterfaceAdapter iface: ip.getInterfaces())
+			iface.removeAttrListener(this);
+	}
+	
+	private void sendToAll(DVMessage.DV dv)
+	throws Exception {
+		for (IPInterfaceAdapter iface: ip.getInterfaces()) {
+			if (iface instanceof IPLoopbackAdapter)
+				continue;
+			DVMessage dvm= new DVMessage();
+			dvm.addDV(dv.dst, dv.metric, null);
+			iface.send(new Datagram(iface.getAddress(), IPAddress.BROADCAST, IP_PROTO_DV, 1, dvm), null);
+		}
+	}
+
+	private DVMessage.DV compute(Map<IPAddress,DVMessage.DV> dvs) {
+		System.out.println("\tcompute");
+		DVMessage.DV bestDV= null;
+		for (IPAddress neighbor: dvs.keySet()) {
+			DVMessage.DV dv= dvs.get(neighbor);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.